### PR TITLE
Fixed chunk writing if the data is bugger than our buffer (#251)

### PR DESCRIPTION
**Pull request by zanker**

### CODE REVIEW

**Comment by zanker** [comment]
https://github.com/httprb/http/issues/251

cc @tarcieri @ixti @sferik 

The failure mode for this is it either times out, or hangs forever. So apparently nobody has ran into this or has and hasn't reported it.

**Comment by tarcieri** [comment]
One related-looking test failure?

Would be nice if we had a better strategy for write buffering in general. I'm pretty sure we run afoul of the pathological case for Nagle where do a small write immediately followed by another write.

**Comment by zanker** [comment]
:( socket mocking. I want to redo the test suite somepoint and pull in the one we use for our internal stuff, since that has a lot of edge case handling already too.

Issue should be fixed

**Comment by tarcieri** [comment]
LGTM :+1:

File: lib/http/request/writer.rb [hunk 1]
```diff
@@ -65,24 +68,35 @@ def send_request_header
...
         end
       end
 
       private
 
+      def write(data)
+        while data.present?
+          length = @socket.write(data)
+          if data.length > length
+            data = data[length..-1]
```
Comment (tarcieri) [inline lib/http/request/writer.rb]: This will allocate every time. I wonder if it would be possible to use `#slice!` here to reduce allocations. Or maybe we just need real `ByteBuffers` :confused:

Comment (zanker) [inline lib/http/request/writer.rb]: We call this using `@body` as is, so we probably can't use `slice!`, unless we call `dup` beforehand.

Would need to benchmark to figure out if the optimization works out.

Comment (tarcieri) [inline lib/http/request/writer.rb]: Since Ruby strings are CoW anyway, this usage is probably efficient
